source/slang/slang-ast-serialize.h

```diff
@@ -7,13 +7,188 @@
+
+The AST node types are generally types derived from the NodeBase. The C++ extractor is used to associate an ASTNodeType with
+every NodeBase type, such that casting is fast and simple and we have a simple integer to uniquely identify those types. The
+extractor also performs another task of associating with the type name all of the fields held in just that type. The definition
+of the fields is stored in an 'x macro' which is in the slang-ast-generated-macro.h file, for example
+
+```
+#define SLANG_FIELDS_ASTNode_DeclRefExpr(_x_, _param_)\
+    _x_(scope, (RefPtr<Scope>), _param_)\
+    _x_(declRef, (DeclRef<Decl>), _param_)\
+    _x_(name, (Name*), _param_)
+``
+
+For the type DeclRefExpr, this holds all of the fields held in just DeclRefExpr in this case `scope`, `declRef` and `name`.
+DeclRefExpr derives from Expr and this might hold other fields and so forth.
+
+The implementation makes a distinction between the 'native' types, the regular C++ in memory types and 'serial' types.
+Each serializable C++ type has an associated 'serial' type - with the distinction that it can be written out and (with perhaps some other data)
+read back in to recreate the C++ type. The serial type can be a C++ type, but is such it can be written and read from disk and still
+represent the same data. 
+
+We need a mechanism to be able to do do a conversion between native and serial types. To make the association we use the template
+
+```
+template <typename T>
+struct ASTSerialTypeInfo;
+```
+
+and specialize it for each native type. The specialization holds
+
+SerialType - The type that will be used to represent the native type
+NativeType - The native typs
+SerialAlignment - A value that holds what kind of alignment the SerialType needs to be serializable (it may be different from SLANG_ALIGN_OF(SerialType)!)
+toSerial - A function that with the help of ASTSerialWriter convert the NativeType into the SerialType
+toNative - A function that with the help of ASTSerialReader convert the SerialType into the NativeType
+
+It is useful to have a structure that holds the type information, so it can be stored. That is achieved with
+
+```
+template <typename T>
+struct ASTSerialGetType;
+```
+
+This template can be specialized for a specific native types - but all it holds is just a function getType, which returns a ASTSerialType*,
+which just holds the information held in the ASTSerialTypeInfo template, but additionally including the size of the SerialType.
+
+So we need to define a specialized ASTSerialTypeInfo for each type that can be a field in a NodeBase derived type. We don't need to define
+anything explicitly for the NodeBase derived types, as we will just generate the layout from the fields. How do we know the fields? We just
+used the macros generated from the C++ extactor.
+
+So first a few things to observe...
+
+1) Some types don't need any conversion to be serializable - int8_t, or float the bits can just be written out and read in (1)
```

Comment:
I assume that we are implicitly following the policy of the rest of the codebase to juse assume little endian has taken over enough of the world that worrying about endian conversion doesn't pay off.

Reply:
For ir serialization - I did nod to endianess in that with that serialized format it is relatively easy to handle endianness. Here it could be supported but would probably want some other template stuff to define a per serialized type function that could handle endianess. That it only needs to work at the field type level (and only with serialized types) - as NodeBase types can just use that field knowledge to flip those types.